Firestore/core/src/timestamp_internal.cc

```diff
@@ -16,12 +16,47 @@
 
 #include "Firestore/core/src/timestamp_internal.h"
 
+#include "Firestore/core/include/firebase/firestore/firestore_errors.h"
 #include "Firestore/core/src/util/hashing.h"
+#include "Firestore/core/src/util/statusor.h"
 
 namespace util = firebase::firestore::util;
 
 namespace firebase {
 
+util::StatusOr<Timestamp> TimestampInternal::FromUntrustedTime(
+    absl::Time time) {
+  // Note `ToUnixSeconds` rounds towards negative infinity, this makes the
+  // `nanos` calculated below always non-negative, meeting protobuf's
+  // requirement.
+  auto seconds = absl::ToUnixSeconds(time);
+  auto nanos = (time - absl::FromUnixSeconds(seconds)) / absl::Nanoseconds(1);
+  return FromUntrustedSecondsAndNanos(seconds, nanos);
+}
+
+util::StatusOr<Timestamp> TimestampInternal::FromUntrustedSecondsAndNanos(
+    int64_t seconds, int32_t nanos) {
+  // The Timestamp ctor will assert if we provide values outside the valid
+  // range. However, since we're decoding, a single corrupt byte could cause
+  // this to occur, so we'll verify the ranges before passing them in since we'd
+  // rather not abort in these situations.
+  if (seconds < TimestampInternal::Min().seconds()) {
+    return util::Status(
+        firestore::Error::kErrorDataLoss,
```

Comment:
Data loss seems like the wrong error by default. It seems like this would be an invalid argument here and then possibly translated to data loss in contexts where we're parsing existing data.

Reply:
Done.